src/ivc/step_folding_circuit.rs

```diff
@@ -0,0 +1,401 @@
+                )?;
+
+                let next_step_i =
+                    gate.add_with_const(&mut region, &assigned_step_i, C::Base::ONE)?;
+
+                let assigned_is_zero_step = gate.is_zero_term(&mut region, assigned_step_i)?;
+
+                let new_U = AssignedRelaxedPlonkInstance::<C>::conditional_select(
+                    &mut region,
+                    &config.main_gate_config,
+                    &U_new_non_base,
+                    &U_new_base,
+                    assigned_is_zero_step.clone(),
+                )?;
+
+                let assigned_input: [_; ARITY] = assigned_z_0
+                    .iter()
+                    .zip_eq(assigned_z_i.iter())
+                    .map(|(z_0, z_i)| {
+                        gate.conditional_select(&mut region, z_0, z_i, &assigned_is_zero_step)
+                    })
+                    .collect::<Result<Vec<_>, _>>()?
+                    .try_into()
+                    .unwrap();
+
+                Ok((next_step_i, new_U, assigned_input))
+            },
+        )?;
+
+        let z_output =
+            self.step_circuit
+                .synthesize_step(config.step_config, layouter, &assigned_input)?;
+
+        let output_hash = layouter.assign_region(
+            || "generate input",
+            |region| {
+                let mut ctx = RegionCtx::new(region, 0);
+
+                let bits = RO::new(
+                    config.main_gate_config.clone(),
+                    self.input.step_pp.ro_constant.clone(),
+                )
+                .absorb_point(WrapValue::from_assigned_point(
+                    &assigned_input_witness.public_params_hash,
+                ))
+                .absorb_base(WrapValue::Assigned(assigned_next_step_i.clone()))
+                .absorb_iter(assigned_z_0.iter())
+                .absorb_iter(z_output.iter().cloned())
+                .absorb_iter(assigned_new_U.iter_wrap_values())
+                .squeeze_n_bits(&mut ctx, NUM_CHALLENGE_BITS)?;
+
+                MainGate::new(config.main_gate_config.clone()).le_bits_to_num(&mut ctx, &bits)
+            },
+        )?;
+
+        layouter.constrain_instance(
+            assigned_input_witness.input_instance[1].0.cell(),
+            config.instance,
+            0,
+        )?;
+        layouter.constrain_instance(output_hash.cell(), config.instance, 1)?;
+
+        Ok(z_output)
+    }
+}
+
+impl<'link, const ARITY: usize, C, SC, RO, const T: usize>
+    StepFoldingCircuit<'link, ARITY, C, SC, RO, T>
+where
+    C: CurveAffine,
+    C::Base: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    C::Scalar: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    SC: StepCircuit<ARITY, C::Base> + Sized,
+    RO: ROCircuitTrait<C::Base, Config = MainGateConfig<T>>,
+{
+    pub fn synthesize(
+        &self,
+        config: StepConfig<ARITY, C::Base, SC, T>,
+        layouter: &mut impl Layouter<C::Base>,
+    ) -> Result<[C::Base; ARITY], SynthesisError> {
```

Comment:
NBD: We can make `StepFoldingCircuit` follow `StepCircuit` trait to return `[Assigned<C::Base>; ARITY]`. However, it's not needed for now until we do parallel ivc folding.

Reply:
We can add a comment to indicate potential future change

Reply:
Look carefully, I implemented `StepCircuit` for this circuit, so it already does what you described

However, we don't need assigned values at the IVC level, so there is this function